Clustering/Hdbscan/src/main/java/org/tribuo/clustering/hdbscan/HdbscanTrainer.java

```diff
@@ -0,0 +1,828 @@
+        for (int i = 0; i < clusterLabels.size(); i++) {
+            Integer clusterLabel = clusterLabels.get(i);
+            Double outlierScore = outlierScoresVector.get(i);
+            if (clusterAssignments.containsKey(clusterLabel)) {
+                clusterAssignments.get(clusterLabel).put(outlierScore, i);
+            }
+            else {
+                TreeMap<Double, Integer> outlierScoreIndexTree = new TreeMap<>();
+                outlierScoreIndexTree.put(outlierScore, i);
+                clusterAssignments.put(clusterLabel, outlierScoreIndexTree);
+            }
+        }
+        return clusterAssignments;
+    }
+
+    /**
+     * Compute the exemplars. These are representative points which are subsets of their clusters and noise points, and
+     * will be used for prediction on unseen data points.
+     *
+     * @param data An array of {@link DenseVector} containing the data.
+     * @param clusterAssignments A map of the cluster labels, and the points assigned to them.
+     * @return A list of {@link ClusterExemplar}s which are used for predictions.
+     */
+    private static List<ClusterExemplar> computeExemplars(SGDVector[] data, Map<Integer, TreeMap<Double, Integer>> clusterAssignments) {
+        List<ClusterExemplar> clusterExemplars = new ArrayList<>();
+        // The formula to calculate the exemplar number. This calculates the number of exemplars to be used for this
+        // configuration. The appropriate number of exemplars is important for prediction. At the time, this
+        // provides a good value for most scenarios.
+        int mumExemplars = (int) Math.sqrt(data.length / 2.0) + clusterAssignments.size();
+
+        for (Entry<Integer, TreeMap<Double, Integer>> e : clusterAssignments.entrySet()) {
+            int clusterLabel = e.getKey();
+            TreeMap<Double, Integer> outlierScoreIndexTree = clusterAssignments.get(clusterLabel);
+            int numExemplarsThisCluster = e.getValue().size() * mumExemplars / data.length;
+            if (clusterLabel != OUTLIER_NOISE_CLUSTER_LABEL) {
+                for (int i = 0; i < numExemplarsThisCluster; i++) {
+                    // Note that for non-outliers, the first node is polled from the tree, which has the lowest outlier
+                    // score out of the remaining points assigned this cluster.
+                    Entry<Double, Integer> entry = outlierScoreIndexTree.pollFirstEntry();
+                    clusterExemplars.add(new ClusterExemplar(clusterLabel, entry.getKey(), data[entry.getValue()]));
+                }
+            }
+            else {
+                for (int i = 0; i < numExemplarsThisCluster; i++) {
+                    // Note that for outliers the last node is polled from the tree, which has the highest outlier score
+                    // out of the remaining points assigned this cluster.
+                    Entry<Double, Integer> entry = outlierScoreIndexTree.pollFirstEntry();
```

Comment:
The comment says this polls the last node, but the code is polling the first node?

Reply:
That is a bug! Thanks for catching this. For harder datasets this subtle issue would really effect predictions.